tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -0,0 +1,331 @@
+                file_info = FileInfo(uri, context)
+            contexts[uri] = file_info
+
+        return (file_info, errors)
+
+    def build_diagnostics(uri):
+        """
+        Parse a file at a particular URI, run the linter rules on the resulting
+        ASTs, and return them as LSP diagnostics.
+        """
+
+        fi, errors = get_context(uri, do_update=True)
+        with fi.context.track_errors() as new_errors:
+            _ = fi.parse_file()
+        errors.extend(new_errors)
+
+        diagnostics = []
+
+        kind_to_severity = {
+            "error": DiagnosticSeverity.Error,
+            "syntax": DiagnosticSeverity.Error,
+            "note": DiagnosticSeverity.Information,
+            "warning": DiagnosticSeverity.Warning,
+        }
+
+        for error in errors:
+            diagnostic = Diagnostic(
+                range=location_to_range(error.location()),
+                message="{}: [{}]: {}".format(
+                    error.kind().capitalize(), error.type(), error.message()
+                ),
+                severity=kind_to_severity[error.kind()],
+            )
+            diagnostics.append(diagnostic)
+        return diagnostics
+
+    # The following functions are handlers for LSP events received by the server.
+
+    @server.feature(TEXT_DOCUMENT_DID_OPEN)
+    @server.feature(TEXT_DOCUMENT_DID_SAVE)
+    async def did_save(
+        ls: LanguageServer,
+        params: Union[DidSaveTextDocumentParams, DidOpenTextDocumentParams],
+    ):
+        text_doc = ls.workspace.get_text_document(params.text_document.uri)
+        diag = build_diagnostics(text_doc.uri)
+        ls.publish_diagnostics(text_doc.uri, diag)
+
+    @server.feature(TEXT_DOCUMENT_DEFINITION)
+    async def get_def(ls: LanguageServer, params: DefinitionParams):
+        text_doc = ls.workspace.get_text_document(params.text_document.uri)
+
+        fi, _ = get_context(text_doc.uri)
+        segments = fi.segments
+
+        idx = (
+            bisect_right(segments, params.position, key=lambda s: s.ident.rng.start) - 1
+        )
+        if idx >= 0 and params.position < segments[idx].ident.rng.end:
+            return segments[idx].resolved_to.get_location()
+
+        return None
+
+    @server.feature(TEXT_DOCUMENT_DOCUMENT_SYMBOL)
+    async def get_sym(ls: LanguageServer, params: DocumentSymbolParams):
+        text_doc = ls.workspace.get_text_document(params.text_document.uri)
+
+        fi, _ = get_context(text_doc.uri)
+
+        # doesn't descend into nested definitions for Functions
+        def preorder_ignore_funcs(node):
+            yield node
+            if isinstance(node, chapel.core.Function):
+                return
+            for child in node:
+                yield from preorder_ignore_funcs(child)
+
+        syms = [
+            si
+            for node, _ in chapel.each_matching(
+                fi.get_asts(), chapel.core.NamedDecl, iterator=preorder_ignore_funcs
```

Comment:
Surprised `iterator=preorder_ignore_funcs` isn't on a new line. It wraps in GitHub.

Reply:
reformatted and fixed